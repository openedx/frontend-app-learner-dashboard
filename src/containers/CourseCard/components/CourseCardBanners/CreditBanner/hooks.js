import { StrictDict } from '../../../../../utils';
import { reduxHooks } from '../../../../../hooks';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = reduxHooks.useCardCreditData(cardId);
  const { supportEmail } = reduxHooks.usePlatformSettingsData();
  if (!credit.isEligible) {
    return null;
  }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
