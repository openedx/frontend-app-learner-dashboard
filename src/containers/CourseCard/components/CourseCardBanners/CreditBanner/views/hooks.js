import React from 'react';

import { StrictDict } from '@src/utils';
import { apiHooks } from '@src/hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest()
      .then((request) => {
        setRequestData(request.data);
      });
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
