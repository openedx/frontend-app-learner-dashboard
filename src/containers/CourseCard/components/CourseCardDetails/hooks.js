import { useIntl } from '@edx/frontend-platform/i18n';
import { utilHooks, reduxHooks } from 'hooks';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const enrollment = reduxHooks.useCardEnrollmentData(cardId);
  const courseRun = reduxHooks.useCardCourseRunData(cardId);
  const formatDate = utilHooks.useFormatDate();
  if (!courseRun.isStarted) {
    if (!courseRun.startDate && !courseRun.advertisedStart) { return null; }
    const startDate = courseRun.advertisedStart ? courseRun.advertisedStart : formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = reduxHooks.useCardProviderData(cardId).name;
  const { courseNumber } = reduxHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = reduxHooks.useCardEntitlementData(cardId);

  const openSessionModal = reduxHooks.useUpdateSelectSessionModalCallback(cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
