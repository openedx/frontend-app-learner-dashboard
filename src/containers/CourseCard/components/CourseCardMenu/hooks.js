import { useKeyedState } from '@edx/react-unit-test-utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

export const stateKeys = {
  isUnenrollConfirmVisible: 'isUnenrollConfirmVisible',
  isEmailSettingsVisible: 'isEmailSettingsVisible',
};

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = useKeyedState(stateKeys.isUnenrollConfirmVisible, false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = useKeyedState(stateKeys.isEmailSettingsVisible, false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(
    track.course.courseOptionsDropdownClicked,
    cardId,
  );
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useOptionVisibility = (cardId) => {
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);

  const shouldShowUnenrollItem = isEnrolled && !isEarned;
  const shouldShowDropdown = (
    shouldShowUnenrollItem
    || isEmailEnabled
    || facebook.isEnabled
    || twitter.isEnabled
  );

  return {
    shouldShowUnenrollItem,
    shouldShowDropdown,
  };
};
