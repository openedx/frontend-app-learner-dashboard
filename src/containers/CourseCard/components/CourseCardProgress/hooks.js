import { useState, useEffect } from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';

import { reduxHooks } from 'hooks';
import messages from './messages';

export const useCourseProgress = ({ cardId }) => {
  const [progress, setProgress] = useState(0);
  const [completeCount, setCompleteCount] = useState(0);
  const [totalCount, setTotalCount] = useState(0);
  const [loading, setLoading] = useState(false);
  const { isEnrolled, hasStarted } = reduxHooks.useCardEnrollmentData(cardId);
  const { courseId } = reduxHooks.useCardCourseRunData(cardId);

  useEffect(() => {
    const fetchProgress = async () => {
      if (!isEnrolled || !hasStarted || !courseId) {
        setProgress(0);
        setCompleteCount(0);
        setTotalCount(0);
        return;
      }

      setLoading(true);
      try {
        const client = getAuthenticatedHttpClient();
        const lmsApiUrl = `${getConfig().LMS_BASE_URL}/api/course_home/v1/progress/${courseId}`;
        const response = await client.get(lmsApiUrl);
        const progressData = response.data;

        let progressPercent = 0;
        let complete = 0;
        let total = 0;

        // Check for completion_summary from course_home API
        if (progressData.completion_summary) {
          const { complete_count, incomplete_count, locked_count } = progressData.completion_summary;
          complete = complete_count;
          total = complete_count + incomplete_count + locked_count;
          if (total > 0) {
            progressPercent = Math.round((complete_count / total) * 100);
          }
        } else if (progressData.completion_percentage !== undefined) {
          progressPercent = progressData.completion_percentage;
        } else if (progressData.progress?.percent !== undefined) {
          progressPercent = Math.round(progressData.progress.percent * 100);
        } else if (progressData.progress_percentage !== undefined) {
          progressPercent = progressData.progress_percentage;
        } else if (progressData.percent !== undefined) {
          progressPercent = progressData.percent;
        }

        setProgress(Math.max(0, Math.min(100, progressPercent)));
        setCompleteCount(complete);
        setTotalCount(total);
      } catch (error) {
        console.warn('Failed to fetch course progress:', error);
        setProgress(0);
        setCompleteCount(0);
        setTotalCount(0);
      } finally {
        setLoading(false);
      }
    };

    fetchProgress();
  }, [cardId, isEnrolled, hasStarted, courseId]);

  return { progress, completeCount, totalCount, loading };
};

export const useProgressVariant = ({ progress }) => {
  if (progress >= 75) return 'success';
  if (progress >= 50) return 'warning';
  return 'info';
};

export const useCardProgressData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { isEnrolled } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    progress, completeCount, totalCount, loading,
  } = useCourseProgress({ cardId });
  const variant = useProgressVariant({ progress });

  return {
    shouldRender: isEnrolled,
    progress,
    completeCount,
    totalCount,
    loading,
    variant,
    progressLabel: formatMessage(messages.courseProgress),
    loadingLabel: formatMessage(messages.loading),
  };
};

export default useCardProgressData;
