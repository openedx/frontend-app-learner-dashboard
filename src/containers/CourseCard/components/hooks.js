import { reduxHooks } from 'hooks';

export const useActionDisabledState = (cardId) => {
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const {
    hasAccess, isAudit, isAuditAccessExpired,
  } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    isEntitlement, isFulfilled, canChange, hasSessions,
  } = reduxHooks.useCardEntitlementData(cardId);

  const { resumeUrl, homeUrl } = reduxHooks.useCardCourseRunData(cardId);

  const disableBeginCourse = !homeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableResumeCourse = !resumeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableViewCourse = !hasAccess || (isAudit && isAuditAccessExpired);
  const disableSelectSession = !isEntitlement || isMasquerading || !hasAccess || (!canChange || !hasSessions);

  const disableCourseTitle = (isEntitlement && !isFulfilled) || disableViewCourse;

  return {
    disableBeginCourse,
    disableResumeCourse,
    disableViewCourse,
    disableSelectSession,
    disableCourseTitle,
  };
};

export default useActionDisabledState;
