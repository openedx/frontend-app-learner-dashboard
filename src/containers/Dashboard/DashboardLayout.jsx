import React from 'react';
import PropTypes from 'prop-types';

import { Container, Col, Row } from '@openedx/paragon';

import WidgetSidebarSlot from 'plugin-slots/WidgetSidebarSlot';

import hooks from './hooks';

export const columnConfig = {
  courseList: {
    withSidebar: {
      lg: { span: 12, offset: 0 },
      xl: { span: 8, offset: 0 },
    },
    noSidebar: {
      lg: { span: 12, offset: 0 },
      xl: { span: 12, offset: 0 },
    },
  },
  sidebar: {
    lg: { span: 12, offset: 0 },
    xl: { span: 4, offset: 0 },
  },
};

export const DashboardLayout = ({ children }) => {
  const {
    isCollapsed,
    sidebarShowing,
  } = hooks.useDashboardLayoutData();

  const courseListColumnProps = sidebarShowing
    ? columnConfig.courseList.withSidebar
    : columnConfig.courseList.noSidebar;

  return (
    <Container fluid size="xl">
      <Row>
        <Col {...courseListColumnProps} className="course-list-column">
          {children}
        </Col>
        <Col {...columnConfig.sidebar} className={['sidebar-column', !isCollapsed && 'not-collapsed']}>
          <WidgetSidebarSlot />
        </Col>
      </Row>
    </Container>
  );
};
DashboardLayout.propTypes = {
  children: PropTypes.node.isRequired,
};

export default DashboardLayout;
