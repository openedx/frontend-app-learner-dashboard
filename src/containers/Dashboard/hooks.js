import React from 'react';
import { useWindowSize, breakpoints } from '@openedx/paragon';
import { useIntl } from '@openedx/frontend-base';
import { apiHooks } from '../../hooks';
import { StrictDict } from '../../utils';
import appMessages from '../../messages';
import * as module from './hooks';

export const state = StrictDict({
  sidebarShowing: (val) => React.useState(val), // eslint-disable-line
});

export const useInitializeDashboard = () => {
  const initialize = apiHooks.useInitializeApp();
  React.useEffect(() => { initialize(); }, []); // eslint-disable-line
};

export const useDashboardMessages = () => {
  const { formatMessage } = useIntl();
  return {
    spinnerScreenReaderText: formatMessage(appMessages.loadingSR),
    pageTitle: formatMessage(appMessages.pageTitle),
  };
};

export const useDashboardLayoutData = () => {
  const { width } = useWindowSize();

  const [sidebarShowing, setSidebarShowing] = module.state.sidebarShowing(true);
  return {
    isDashboardCollapsed: width < breakpoints.large.maxWidth,
    sidebarShowing,
    setSidebarShowing,
  };
};

export default {
  useDashboardLayoutData,
  useInitializeDashboard,
  useDashboardMessages,
};
