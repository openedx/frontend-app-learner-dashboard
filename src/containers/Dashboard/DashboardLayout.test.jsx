import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { Col, Row } from '@openedx/paragon';

import hooks from './hooks';
import DashboardLayout, { columnConfig } from './DashboardLayout';

jest.mock('./hooks', () => ({
  useDashboardLayoutData: jest.fn(),
}));

const hookProps = {
  isCollapsed: true,
  sidebarShowing: false,
  setSidebarShowing: jest.fn().mockName('hooks.setSidebarShowing'),
};
hooks.useDashboardLayoutData.mockReturnValue(hookProps);

const children = 'test-children';

let el;
describe('DashboardLayout', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    el = shallow(<DashboardLayout>{children}</DashboardLayout>);
  });

  const testColumns = () => {
    it('loads courseList and sidebar column layout', () => {
      const columns = el.instance.findByType(Row)[0].findByType(Col);
      Object.keys(columnConfig.sidebar).forEach(size => {
        expect(columns[1].props[size]).toEqual(columnConfig.sidebar[size]);
      });
    });
    it('displays children in first column', () => {
      const columns = el.instance.findByType(Row)[0].findByType(Col);
      expect(columns[0].children).not.toHaveLength(0);
    });
    it('displays WidgetSidebarSlot in second column', () => {
      const columns = el.instance.findByType(Row)[0].findByType(Col);
      expect(columns[1].findByType('WidgetSidebarSlot')).toHaveLength(1);
    });
  };
  const testSidebarLayout = () => {
    it('displays withSidebar width for course list column', () => {
      const columns = el.instance.findByType(Row)[0].findByType(Col);
      Object.keys(columnConfig.courseList.withSidebar).forEach(size => {
        expect(columns[0].props[size]).toEqual(columnConfig.courseList.withSidebar[size]);
      });
    });
  };
  const testNoSidebarLayout = () => {
    it('displays noSidebar width for course list column', () => {
      const columns = el.instance.findByType(Row)[0].findByType(Col);
      Object.keys(columnConfig.courseList.noSidebar).forEach(size => {
        expect(columns[0].props[size]).toEqual(columnConfig.courseList.noSidebar[size]);
      });
    });
  };
  const testSnapshot = () => {
    test('snapshot', () => {
      expect(el.snapshot).toMatchSnapshot();
    });
  };
  describe('collapsed', () => {
    describe('sidebar showing', () => {
      beforeEach(() => {
        hooks.useDashboardLayoutData.mockReturnValueOnce({ ...hookProps, sidebarShowing: true });
      });
      testColumns();
      testSnapshot();
      testSidebarLayout();
    });
    describe('sidebar not showing', () => {
      testColumns();
      testSnapshot();
      testNoSidebarLayout();
    });
    it('does not show spacer component above widget sidebar', () => {
      const columns = el.instance.findByType(Col);
      expect(columns[1].findByType('h2').length).toEqual(0);
    });
  });

  describe('not collapsed', () => {
    const testWidgetSpacing = () => {
      it('shows not-collapsed class on widget sidebar', () => {
        const columns = el.instance.findByType(Col);
        expect(columns[1].props.className).toContain('not-collapsed');
      });
    };
    describe('sidebar showing', () => {
      beforeEach(() => {
        hooks.useDashboardLayoutData.mockReturnValueOnce({
          ...hookProps,
          isCollapsed: false,
          sidebarShowing: true,
        });
      });
      testColumns();
      testSnapshot();
      testSidebarLayout();
      testWidgetSpacing();
    });
    describe('sidebar not showing', () => {
      beforeEach(() => {
        hooks.useDashboardLayoutData.mockReturnValueOnce({ ...hookProps, isCollapsed: false });
      });
      testColumns();
      testSnapshot();
      testNoSidebarLayout();
      testWidgetSpacing();
    });
  });
});
