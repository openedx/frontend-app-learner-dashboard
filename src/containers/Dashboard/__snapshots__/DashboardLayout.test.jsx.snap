// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardLayout collapsed sidebar not showing snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 12,
        }
      }
    >
      test-children
    </Col>
    <Col
      className={
        [
          "sidebar-column",
          false,
        ]
      }
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <WidgetSidebarSlot />
    </Col>
  </Row>
</Container>
`;

exports[`DashboardLayout collapsed sidebar showing snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 8,
        }
      }
    >
      test-children
    </Col>
    <Col
      className={
        [
          "sidebar-column",
          false,
        ]
      }
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <WidgetSidebarSlot />
    </Col>
  </Row>
</Container>
`;

exports[`DashboardLayout not collapsed sidebar not showing snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 12,
        }
      }
    >
      test-children
    </Col>
    <Col
      className={
        [
          "sidebar-column",
          "not-collapsed",
        ]
      }
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <WidgetSidebarSlot />
    </Col>
  </Row>
</Container>
`;

exports[`DashboardLayout not collapsed sidebar showing snapshot 1`] = `
<Container
  fluid={true}
  size="xl"
>
  <Row>
    <Col
      className="course-list-column"
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 8,
        }
      }
    >
      test-children
    </Col>
    <Col
      className={
        [
          "sidebar-column",
          "not-collapsed",
        ]
      }
      lg={
        {
          "offset": 0,
          "span": 12,
        }
      }
      xl={
        {
          "offset": 0,
          "span": 4,
        }
      }
    >
      <WidgetSidebarSlot />
    </Col>
  </Row>
</Container>
`;
